STM32F405RGTx/Core/Src/rtos_threads/state_machine.cpp

```diff
@@ -0,0 +1,472 @@
+    CANFrame tx_frame = CANFrame_init(MOTOR_CONTROLLER_SEVERITY_CODE.id);
+    CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_SEVERITY_CODE, 0x01);
+    CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_ERROR_CODE, mc_error_code);
+    if (CANBus_put_frame(&tx_frame) != HAL_OK) {
+        Error_Handler();
+    }
+    /* After everything is merged
+    Put all the MOSFETs into HIGH-Z mode using the drv8323 drivers
+        . The DRV drivers aren't merged in right now, so just put a comment to remind yourself to put that in once it does get merged
+    (Probably) read the fault codes from the DRV
+    Disable the PID, SVPWM, and VHz threads
+    */
+    // Receive CAN frame
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        if (state_id == RESTING) {
+            return Idle;
+        }
+        else {
+            return NormalDangerFault;
+        }
+    }
+    return NormalDangerFault;
+}
+
+State_t StateMachineThread::SevereDangerFaultEvent(void)
+{
+
+    // TODO: Make list of the threads that need to be turned on or turned off
+        // On: Measurements Thread, CANThread
+        // Off: PIDThread, VHzThread, SVPWMThread
+    MeasurementsThread::resumeMeasurements();
+    /* After everything is merged
+    Put all the MOSFETs into HIGH-Z mode using the drv8323 drivers
+        . The DRV drivers aren't merged in right now, so just put a comment to remind yourself to put that in once it does get merged
+    (Probably) read the fault codes from the DRV
+    Disable the PID, SVPWM, and VHz threads
+    */
+
+    // Report fault on CAN
+    CANFrame tx_frame = CANFrame_init(MOTOR_CONTROLLER_SEVERITY_CODE.id);
+    CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_SEVERITY_CODE, 0x00);
+    CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_ERROR_CODE, mc_error_code);
+    if (CANBus_put_frame(&tx_frame) != HAL_OK) {
+        Error_Handler();
+    }
+
+    // Receive CAN frame
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        if (state_id == RESTING) {
+            return Idle;
+        }
+        else {
+            return NormalDangerFault;
+        }
+    }
+    return NormalDangerFault;
+}
+
+State_t StateMachineThread::NoFaultEvent() {
+    return NoFault;
+}
+
+void StateMachineThread::initialize() {
+    thread = RTOSThread(
+        "state_machine_thread",
+        1024 * 3,
+        osPriorityNormal,
+        runStateMachine);
+}
+
+void StateMachineThread::runStateMachine(void *argument)
+{
+    start_timers();
+    while (1)
+    {
+        CurrentState = NewState;
+        switch (NewState)
+        {
+        case Initialize:
+            NewState = InitializeEvent();
+            break;
+
+        case InitializeFault:
+            red = 50.0;
+            green = 0.0;
+            blue = 0.0;
+            flash = false;
```

Comment:
Maybe add a function like `_ConfigureLED(uint8_t r, uint8_t g, uint8_t b, uint8_t blink)` and have it set the values so this can be replaced with something like `_ConfigureLED(50, 0, 0, 0);`